File: examples/llm-routing/e2e-tests/go/main.go
//go:build !loadtest

// Community LLM Provider E2E Tests using Go SDK
// Tests governed LLM access through AxonFlow Agent
//
// VALIDATION: This example exits with code 1 if any assertion fails.
package main

import (
	"fmt"
	"os"

	axonflow "github.com/getaxonflow/axonflow-sdk-go/v9"
)

var failures []string

func assertCheck(condition bool, message string) {
	if condition {
		fmt.Printf("   PASS: %s\n", message)
	} else {
		fmt.Printf("   FAIL: %s\n", message)
		failures = append(failures, message)
	}
}

func main() {
	// Create client - SDK talks to Agent which routes to Orchestrator
	agentURL := os.Getenv("AXONFLOW_ENDPOINT")
	if agentURL == "" {
		agentURL = os.Getenv("AGENT_URL")
	}
	if agentURL == "" {
		agentURL = "http://localhost:8080"
	}

	userToken := os.Getenv("AXONFLOW_USER_TOKEN")
	if userToken == "" {
		userToken = "test-user@example.com"
	}

	client := axonflow.NewClient(axonflow.AxonFlowConfig{
		Endpoint:     agentURL,
		ClientID:     os.Getenv("AXONFLOW_CLIENT_ID"),
		ClientSecret: os.Getenv("AXONFLOW_CLIENT_SECRET"),
	})

	fmt.Println("=== Community LLM Provider Tests (Go SDK) ===")
	fmt.Printf("Agent URL: %s\n\n", agentURL)

	// Test 1: Health check
	fmt.Println("Test 1: Agent health check")
	if err := client.HealthCheck(); err != nil {
		fmt.Printf("  Health check failed: %v\n", err)
		assertCheck(false, "Agent health check passed")
	} else {
		fmt.Println("  Agent is healthy")
		assertCheck(true, "Agent health check passed")
	}
	fmt.Println()

	// Test 2: Execute query with OpenAI preference
	fmt.Println("Test 2: Per-request selection - OpenAI")
	resp, err := client.ProxyLLMCall(
		userToken,
		"Say hello in 3 words",
		"chat",
		map[string]interface{}{"provider": "openai"},
	)
	if err != nil {
		fmt.Printf("  Error: %v\n", err)
		assertCheck(false, "OpenAI provider query succeeded")
	} else {
		fmt.Printf("  Success: %v, Response: %s\n", resp.Success, truncate(fmt.Sprintf("%v", resp.Data), 50))
		assertCheck(resp.Success || resp.Data != nil, "OpenAI provider query succeeded")
	}
	fmt.Println()

	// Test 3: Execute query with Anthropic preference
	fmt.Println("Test 3: Per-request selection - Anthropic")
	resp, err = client.ProxyLLMCall(
		userToken,
		"Say hello in 3 words",
		"chat",
		map[string]interface{}{"provider": "anthropic"},
	)
	if err != nil {
		fmt.Printf("  Error: %v\n", err)
		assertCheck(false, "Anthropic provider query succeeded")
	} else {
		fmt.Printf("  Success: %v, Response: %s\n", resp.Success, truncate(fmt.Sprintf("%v", resp.Data), 50))
		assertCheck(resp.Success || resp.Data != nil, "Anthropic provider query succeeded")
	}
	fmt.Println()

	// Test 4: Execute query with Gemini preference
	fmt.Println("Test 4: Per-request selection - Gemini")
	resp, err = client.ProxyLLMCall(
		userToken,
		"Say hello in 3 words",
		"chat",
		map[string]interface{}{"provider": "gemini"},
	)
	if err != nil {
		fmt.Printf("  Error: %v\n", err)
		assertCheck(false, "Gemini provider query succeeded")
	} else {
		fmt.Printf("  Success: %v, Response: %s\n", resp.Success, truncate(fmt.Sprintf("%v", resp.Data), 50))
		assertCheck(resp.Success || resp.Data != nil, "Gemini provider query succeeded")
	}
	fmt.Println()

	// Test 5: Weighted routing (no provider preference)
	fmt.Println("Test 5: Weighted routing distribution (5 queries)")
	successCount := 0
	for i := 0; i < 5; i++ {
		resp, err = client.ProxyLLMCall(
			userToken,
			"Hello",
			"chat",
			nil,
		)
		if err != nil {
			fmt.Printf("  Query %d: Error - %v\n", i+1, err)
		} else {
			fmt.Printf("  Query %d: Success\n", i+1)
			successCount++
		}
	}
	assertCheck(successCount > 0, "At least one weighted routing query succeeded")
	fmt.Println()

	fmt.Println("=== Tests Complete ===")

	if len(failures) > 0 {
		fmt.Printf("\nFAILED: %d assertions failed\n", len(failures))
		for _, f := range failures {
			fmt.Printf("  - %s\n", f)
		}
		os.Exit(1)
	}
	fmt.Println("ALL ASSERTIONS PASSED - LLM routing E2E tests verified!")
}

func truncate(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}
	return s[:maxLen] + "..."
}

File: examples/llm-routing/e2e-tests/typescript/main.ts
/**
 * Community LLM Provider E2E Tests using TypeScript SDK
 *
 * VALIDATION: This example exits with code 1 if any assertion fails.
 */
import { AxonFlow } from "@axonflow/sdk";

const failures: string[] = [];

function assertCheck(condition: boolean, message: string): void {
  if (condition) {
    console.log(`   ✓ PASS: ${message}`);
  } else {
    console.log(`   ❌ FAIL: ${message}`);
    failures.push(message);
  }
}

async function main() {
  const endpoint = process.env.AXONFLOW_AGENT_URL || "http://localhost:8080";
  const client = new AxonFlow({
    endpoint,
    clientId: process.env.AXONFLOW_CLIENT_ID || "demo",
    clientSecret: process.env.AXONFLOW_CLIENT_SECRET || "demo-secret",
  });

  console.log("=== Community LLM Provider Tests (TypeScript SDK) ===");
  console.log(`Target: ${endpoint}\n`);

  // Test 1: List providers
  console.log("Test 1: List providers");
  try {
    const providers = await client.listProviders();
    for (const p of providers) {
      console.log(`  - ${p.name} (${p.type}): ${p.health?.status ?? "unknown"}`);
    }
    assertCheck(Array.isArray(providers), "Providers response is an array");
    assertCheck(providers.length > 0, `At least one provider returned (got ${providers.length})`);
  } catch (e) {
    console.log(`  Failed: ${e}`);
    failures.push("Test 1: listProviders failed");
  }
  console.log();

  // Test 2-4: Per-request provider selection.
  // Note: ExecuteQueryResponse currently returns policyInfo + budgetInfo
  // but does not surface providerInfo on the response. We assert the call
  // succeeded and a non-empty response was produced; per-provider tracking
  // is verified at the platform/Prometheus layer.
  const providersToTest = ["openai", "anthropic", "gemini"];
  for (let idx = 0; idx < providersToTest.length; idx++) {
    const provider = providersToTest[idx];
    console.log(`Test ${idx + 2}: Per-request selection - ${provider}`);
    try {
      const resp = await client.proxyLLMCall({
        userToken: "test-user",
        query: "Say hello in 3 words",
        requestType: "chat",
        context: { provider },
      });
      const text =
        resp.data && typeof resp.data === "object"
          ? (resp.data as { data?: string }).data
          : undefined;
      console.log(`  Response: ${truncate(text ?? "", 50)}`);
      assertCheck(resp.success === true, `${provider} request reports success`);
      assertCheck(
        typeof text === "string" && text.length > 0,
        `${provider} response data is non-empty`
      );
    } catch (e) {
      console.log(`  Failed: ${e}`);
      failures.push(`${provider} request failed`);
    }
    console.log();
  }

  // Test 5: Repeated default-routing requests succeed
  console.log("Test 5: Default-route distribution (5 requests)");
  let successCount = 0;
  for (let i = 0; i < 5; i++) {
    try {
      const resp = await client.proxyLLMCall({
        userToken: "test-user",
        query: "Hello",
        requestType: "chat",
      });
      if (resp.success) {
        successCount++;
        console.log(`  Request ${i + 1}: success`);
      } else {
        console.log(`  Request ${i + 1}: not-success`);
      }
    } catch (e) {
      console.log(`  Request ${i + 1}: failed (${e})`);
    }
  }
  assertCheck(
    successCount >= 3,
    `At least 3/5 default-route requests succeeded (got ${successCount})`
  );
  console.log();

  console.log("=== Tests Complete ===");
  console.log();
  if (failures.length > 0) {
    console.log(`FAILED: ${failures.length} assertion(s) failed:`);
    failures.forEach((f) => console.log(`  - ${f}`));
  } else {
    console.log("All assertions passed!");
  }
  process.exit(failures.length > 0 ? 1 : 0);
}

function truncate(s: string, maxLen: number): string {
  return s.length <= maxLen ? s : s.substring(0, maxLen) + "...";
}

main().catch((err) => {
  console.error("Unexpected error:", err);
  process.exit(1);
});
